README.rst
===============
Zookeeper Tools
===============

``zktools`` is a package of tools implementing higher level constructs using
`Apache Zookeeper`_.

It currently provides:

* ``Configuration`` - Zookeeper Configuration Helpers
  to store and load configuration information stored
  in Zookeeper nodes.
* ``Locks`` - A Zookeeper lock with support for
  non-blocking acquire, modeled on Python's Lock objects that also includes a
  `Revocable Shared Locks with Freaking Laser Beams` described in the
  `Zookeeper Recipe's
  <http://zookeeper.apache.org/doc/current/recipes.html#sc_recoverableSharedLocks>`_.

See `the full docs`_ for more  information.

License
=======

``zktools`` is offered under the MPL license.

Authors
=======

``zktools`` is made available by the `Mozilla Foundation`.

.. _Apache Zookeeper: http://zookeeper.apache.org/
.. _the full docs: http://zktools.rtfd.org/


CHANGES.rst
Changelog
=========

0.3 (unreleased)
----------------

Changes
*******

- Use the Zookeeper-provided mtime of a node as the last_modified
  attribute, instead of client specific time.time()
- Use GUID from new Zookeeper recipe to handle connection loss which
  still results in a created node.
- Add safe-call to ensure that commands run reliably in the face of a
  connection loss exception during their call, and have lock code run
  in a separate thread to ensure it doesn't deadlock the ZK event
  thread.
- Removed zc-zookeeper-static requirement as some OS distributions include
  the python zookeeper binding as a system package.

Bugfixes
********

- Fix ZkNode issues with dead-locks due to having locking code inside
  watch callbacks.

Backward Incompatibilities
**************************

- Removed ZkNodeDict which establishes a questionable amount of watches on
  the server. Setting too many watches has multiple drawbacks, larger values
  should be stored as JSON in a single node.

0.2.1 (02/16/2012)
------------------

- Fixed packaging bug.


0.2 (02/03/2012)
----------------

Changes
*******

- Added context manager return to lock to allow use of the 'with'
  statement.
- Refactored to use zc.zk ZooKeeper library for higher level Zookeeper
  abstraction with automatic watch re-establishment.

Features
********

- Node object to retrieve ZNode data from Zookeeper and keep it up
  to date.
- Node objects can have data and children subscribers.
- NodeDict object that maps a shallow tree (one level of children)
  into a dict-like object.

Backward Incompatibilities
**************************

- SharedZkLock has been refactored into ZkWriteLock and ZkReadLock.
- ``revoked`` is a property of Locks, not a method.
- ZkConnection is gone, lock objects, ZkNode, and ZkNodeDict all expect
  zc.zk ZooKeeper instances.


0.1 (11/22/2011)
----------------

Features
********

- Lock implementation, with revokable shared locks.
- Zookeeper connection object with automatic reconnect.


zktools/tests/__init__.py
import unittest

__all__ = ['TestBase']


connection = []


class TestBase(unittest.TestCase):
    @property
    def conn(self):
        from zc.zk import ZooKeeper
        if connection:
            conn = connection[0]
        else:
            conn = ZooKeeper()
            connection.append(conn)
        return conn


zktools/tests/test_node.py
import datetime
import decimal
import time

from nose.tools import eq_
from nose.tools import raises


from zktools.tests import TestBase


class TestNode(TestBase):
    def makeOne(self, *args, **kwargs):
        from zktools.node import ZkNode
        return ZkNode(self.conn, *args, **kwargs)

    def setUp(self):
        if self.conn.exists('/zkTestNode'):
            self.conn.delete_recursive('/zkTestNode')

    def testSetDefaultValue(self):
        n = self.makeOne('/zkTestNode', 234)
        eq_(n.value, decimal.Decimal('234'))

    def testUpdateValue(self):
        n1 = self.makeOne('/zkTestNode')
        n2 = self.makeOne('/zkTestNode')

        eq_(n1.value, n2.value)
        eq_(n1.last_modified, n2.last_modified)

        old_modified = n1.last_modified
        n1.value = 942

        # It can take a fraction of a second on some machines on occasion
        # for the other value to update
        time.sleep(0.1)
        eq_(n1.value, n2.value)
        eq_(n1.last_modified, n2.last_modified)
        self.assertTrue(n1.last_modified > old_modified)

    def testJsonValue(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = dict(alpha=203)
        time.sleep(0.1)
        eq_(n1.value['alpha'], 203)
        n2 = self.makeOne('/zkTestNode', use_json=True)
        time.sleep(0.1)
        eq_(n2.value, n1.value)

    def testBadJson(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = '{[sasdfasdfsd}'
        n2 = self.makeOne('/zkTestNode', use_json=True)
        time.sleep(0.1)
        eq_(n2.value, n1.value)

    def testNormalString(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = 'fred'
        n2 = self.makeOne('/zkTestNode', use_json=True)
        time.sleep(0.1)
        eq_(n2.value, n1.value)

    def testFloat(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = 492.23
        n2 = self.makeOne('/zkTestNode', use_json=True)
        time.sleep(0.1)
        eq_(n2.value, n1.value)

    def testDates(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = datetime.datetime.today()
        n2 = self.makeOne('/zkTestNode', use_json=True)
        time.sleep(0.1)
        eq_(n2.value, n1.value)

        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = datetime.date.today()
        n2 = self.makeOne('/zkTestNode', use_json=True)
        time.sleep(0.1)
        eq_(n2.value, n1.value)

    def testReload(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = now = datetime.datetime.today()
        time.sleep(0.1)
        n1._reload = True
        eq_(n1.value, now)


zktools/tests/test_config.py
import datetime
import decimal
import time

from nose.tools import eq_


from zktools.tests import TestBase


class TestConfig(TestBase):
    def makeOne(self, *args, **kwargs):
        from zktools.configuration import ZkNode
        return ZkNode(self.conn, *args, **kwargs)

    def setUp(self):
        n = self.makeOne('/zkTestNode')
        n._zk.delete('/zkTestNode')

    def testSetDefaultValue(self):
        n = self.makeOne('/zkTestNode', 234)
        eq_(n.value, decimal.Decimal('234'))

    def testUpdateValue(self):
        n1 = self.makeOne('/zkTestNode')
        n2 = self.makeOne('/zkTestNode')

        eq_(n1.value, n2.value)

        n1.value = 942

        # It can take a fraction of a second on some machines on occasion
        # for the other value to update
        time.sleep(0.1)
        eq_(n1.value, n2.value)

    def testJsonValue(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = dict(alpha=203)
        eq_(n1.value['alpha'], 203)
        n2 = self.makeOne('/zkTestNode', use_json=True)
        eq_(n2.value, n1.value)

    def testBadJson(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = '{[sasdfasdfsd}'
        n2 = self.makeOne('/zkTestNode', use_json=True)
        eq_(n2.value, n1.value)

    def testNormalString(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = 'fred'
        n2 = self.makeOne('/zkTestNode', use_json=True)
        eq_(n2.value, n1.value)

    def testFloat(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = 492.23
        n2 = self.makeOne('/zkTestNode', use_json=True)
        eq_(n2.value, n1.value)

    def testDates(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = datetime.datetime.today()
        n2 = self.makeOne('/zkTestNode', use_json=True)
        eq_(n2.value, n1.value)

        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = datetime.date.today()
        n2 = self.makeOne('/zkTestNode', use_json=True)
        eq_(n2.value, n1.value)

    def testReload(self):
        n1 = self.makeOne('/zkTestNode', use_json=True)
        n1.value = now = datetime.datetime.today()
        n1._reload = True
        eq_(n1.value, now)


zktools/node.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""Zookeeper Nodes

This module provides a :class:`ZkNode` object which can represent a single
node from Zookeeper. It can reflect a single value, or a JSON serialized
value.

"""
import datetime
import decimal
import json
import re
import threading

import zookeeper

ZOO_OPEN_ACL_UNSAFE = dict(perms=zookeeper.PERM_ALL, scheme='world',
                           id='anyone')


CONVERSIONS = {
    re.compile(r'^\d+\.\d+$'): decimal.Decimal,
    re.compile(r'^\d+$'): int,
    re.compile(r'^true$', re.IGNORECASE): lambda x: True,
    re.compile(r'^false$', re.IGNORECASE): lambda x: False,
    re.compile(r'^None$', re.IGNORECASE): lambda x: None,
    re.compile(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z$'):
       lambda x: datetime.datetime.strptime(x, '%Y-%m-%dT%H:%M:%S.%fZ'),
    re.compile(r'^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+Z$'):
       lambda x: datetime.datetime.strptime(x, '%Y-%m-%d %H:%M:%S.%fZ'),
    re.compile(r'^\d{4}-\d{2}-\d{2}$'):
       lambda x: datetime.datetime.strptime(x, '%Y-%m-%d'),
}

JSON_REGEX = re.compile(r'^[\{\[].*[\}\]]$')

# Sad fix for http://bugs.python.org/issue7980
# We import and use strptime here to ensure the whole chain is fully
# imported before threading is likely to occur which remedies the bug
assert datetime.datetime.strptime('2004-02-03T12:10:32.4Z',
                                  '%Y-%m-%dT%H:%M:%S.%fZ')


def _load_value(value, use_json=False):
    """Convert a saved value to the best Python match"""
    if use_json and JSON_REGEX.match(value):
        try:
            return json.loads(value)
        except ValueError:
            return value
    for regex, convert in CONVERSIONS.iteritems():
        if regex.match(value):
            return convert(value)
    return value


def _save_value(value, use_json=False):
    """Convert a Python object to the best string repr"""
    # Float is all we care about, as we lose float precision
    # when calling str on it
    if isinstance(value, float):
        return repr(value)
    elif isinstance(value, datetime.datetime):
        return value.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
    elif isinstance(value, datetime.date):
        return value.strftime('%Y-%m-%d')
    elif use_json and isinstance(value, (dict, list)):
        return json.dumps(value)
    else:
        return str(value)


class ZkNode(object):
    """Zookeeper Node

    This object provides access to a single node for updating the value
    that can also track changes to the value from Zookeeper. Functions
    can be subscribed to changes in the node's value and/or changes in
    the node's children (nodes under it being created/removed).

    The value of the node is coerced into an appropriate Python
    object when loaded, current supported conversions::

        Numbers with decimals      -> Decimal
        Numbers without decimals   -> Int
        true/false                 -> Bool
        none                       -> None
        ISO 8601 Date and Datetime -> date or datetime

    And optionally, with use_json::

        JSON string                -> dict/list

    .. note::

        The JSON determination is extremely lax, if it is a string that
        starts and ends with brackets or curly marks, it is assumed to
        be a JSON object and will be coerced if possible. If coercion
        fails, the string will be returned as is.

    Example::

        from zc.zk import ZooKeeper
        from zktools.node import ZkNode

        conn = ZooKeeper()
        node = ZkNode(conn, '/some/config/node')

        # prints out the current value, defaults to None
        print node.value

        # Set the value in zookeeper
        node.value = 483.24

        # Show the children of the node
        print list(node.children)

        # Subscribe a function to be called when the node's
        # children change (note this will be called immediately
        # with a zc.zk Children instance)
        @node.children
        def my_function(children):
            # do something with node.children or prior_children

    The default behavior is to track changes to the node, so that
    the ``value`` attribute always reflects the node's value in
    Zookeeper. Additional subscriber functions are called when the
    Zookeeper event watch is triggered and are run in a separate
    event thread. Depending on how fast the value/children are
    changing the subscriber functions may run consecutively and
    could miss intermediate values.

    Return values of subscriber functions are ignored.

    .. warning::

        **Do not delete nodes that are in use**, there intentionally is
        no code to handle such conditions as it creates overly complex
        scenarios both for ZkNode and for application code using it.

    """
    def __init__(self, connection, path, default=None, use_json=False,
                 permission=ZOO_OPEN_ACL_UNSAFE, create_mode=0):
        """Create a Zookeeper Node

        Creating a ZkNode by default attempts to load the value, and
        if it is not found will automatically create a blank string as
        the value.

        The last time a :class:`ZkNode` has been modified either by
        the user or due to a Zookeeper update is recorded as the
        :obj:`ZkNode.last_modified` attribute, as a long in
            milliseconds from epoch.

        :param connection: Zookeeper connection object
        :type connection: zc.zk Zookeeper instance
        :param path: Path to the Zookeeper node
        :type path: str
        :param default: A default value if the node is being created
        :param use_json: Whether values that look like a JSON object should
                         be deserialized, and dicts/lists saved as JSON.
        :type use_json: bool
        :param permission: Node permission to use if the node is being
                           created.
        :type permission: dict
        :param create_mode: Persistent or ephemeral creation mode
        :type create_mode: int
        """
        self._zk = connection
        self._path = path
        self._cv = threading.Condition()
        self._use_json = use_json
        self._value = None
        self._reload_data = False

        if not connection.exists(path):
            self._zk.create(self._path,
                            _save_value(default, use_json=use_json),
                            [permission], create_mode)
        self._load()

    def _node_watcher(self, handle, type, state, path):
        """Watch a node for updates"""
        if type == zookeeper.CHANGED_EVENT:
            data = self._zk.get(self._path, self._node_watcher)
            self.last_modified = data[1][u'mtime']
            self._value = _load_value(data[0], use_json=self._use_json)
        elif type in (zookeeper.EXPIRED_SESSION_STATE,
                      zookeeper.AUTH_FAILED_STATE):
            self._reload_data = True

    def _load(self):
        """Load data from the node, and coerce as necessary"""
        data = self._zk.get(self._path, self._node_watcher)
        self.last_modified = data[1][u'mtime']
        self._value = _load_value(data[0], use_json=self._use_json)

    @property
    def value(self):
        """Returns the current value

        If the Zookeeper session expired, it will be reconnected and
        the value reloaded.

        """
        if self._reload_data:
            self._load()
            self._reload = False
        return self._value

    @value.setter
    def value(self, value):
        """Set the value to a new one

        :param value: The value of the node
        :type value: Any str'able object

        """
        self._value = val = _save_value(value, use_json=self._use_json)
        self._zk.set(self._path, val)

    @property
    def connected(self):
        """Indicate whether a connection to Zookeeper exists"""
        return self._zk.connected


zktools/util.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""Utility functions for Zookeeper"""
import uuid
from functools import wraps
from threading import Thread

import zookeeper


def safe_call(zk, func, *args, **kwargs):
    """Safely call a function while handling connection loss

    .. note::

        This function merely retries the query until an active
        Zookeeper session is available that returns without throwing
        a ConnectionLoss Exception. When creating a node, this could
        result in a NodeAlreadyExists exception because the create
        ran twice.

    """
    while 1:
        try:
            return getattr(zk, func)(*args, **kwargs)
        except (zookeeper.ClosingException,
                zookeeper.ConnectionLossException,
                zookeeper.OperationTimeoutException):
            zk.connected.wait()


def safe_create_ephemeral_sequence(zk, name, data, acl):
    """Safely creates an ephemeral sequence node using a UUID prefix

    This function properly handles zookeeper ConnectionLoss exceptions
    and determines after waiting for reconnection whether it was created
    successfully.

    :param zk: Zookeeper instance
    :param name: Name of the node, it will be prefixed by the UUID
    :param data: Data to set on the node
    :param acl: ACL to set for the node
    :returns: Name of the created node

    The name will be split so that its prefixed by the UUID and
    suffixed by the sequence.

    Example:

    .. code-block:: pycon

        >>> safe_create_ephemeral_sequence(zk, '/path/to/node/myname',
                                           [ZOO_OPEN_ACL_UNSAFE])
        '/path/to/node/dfad3fa294d745e499d883b0a38bbc93-myname-0001'

    """
    prefix = uuid.uuid4().hex + '-'
    path, node_name = name.rsplit('/', 1)
    node_name = '/'.join([path, prefix + node_name + '-'])

    while 1:
        try:
            return zk.create(node_name, data, acl,
                             zookeeper.EPHEMERAL | zookeeper.SEQUENCE)
        except (zookeeper.ClosingException,
                zookeeper.ConnectionLossException,
                zookeeper.OperationTimeoutException):
            # Check children to see if the node was created
            children = safe_call(zk, 'get_children', path)
            created = [x for x in children if x.startswith(prefix)]
            if created:
                return '/'.join([path, created[0]])
            # We've verified the create failed, retry
            continue


def threaded(func):
    """Decorator to run a function in a separate thread

    :param func: Function to run in the other thread
    :returns: :class:`Thread` reference

    Example::

        @threaded
        def task1():
            do_something

        @threaded
        def task2():
            do_something_too

        t1 = task1()
        t2 = task2()
        ...
        t1.join()
        t2.join()

    """
    @wraps(func)
    def threaded_func(*args, **kwargs):
        func_hl = Thread(target=func, args=args, kwargs=kwargs)
        func_hl.start()
        return func_hl
    return threaded_func


zktools/connection.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""Zookeeper Connection Classes"""
import time
import threading

import zookeeper


class ZkConnection(object):
    """Zookeeper Connection object"""
    def __init__(self, host="localhost:2181", connect_timeout=10,
                 session_timeout=10 * 1000, reconnect_timeout=10 * 1000,
                 debug=False):
        """Create a connection object, capable of automatically
        reconnecting when the connection is lost.

        In the event the connection is lost, the zookeeper command
        will block until the connection is available again, or the
        ``reconnect_timeout`` is reached. The latter will raise an
        Exception about the reconnect period expiring.

        Example::

            conn = ZkConnection()
            conn.connect()

            node = conn.create(
                "/my/node", "a value",
                [{"perms": 0x1f, "scheme": "world", "id": "anyone"}],
                0)

        .. note::

            ZkConnection ensures that all zookeeper functions use the
            same zookeeper handle that ZkConnection has, so it does
            not need to be passed in as the first argument.

        :param host: A valid zookeeper host string
        :param connect_timeout: Timeout for connecting to zookeeper
        :type connect_timeout: int
        :param session_timeout: Timeout for the zookeeper session
        :type session_timeout: int
        :param reconnect_timeout: How many seconds to attempt to
                                  reconnect before giving up. Set to 0
                                  to avoid reconnect attempts.
        :type reconnect_timeout: int

        """
        self.connected = False
        self._establish_connection = False
        self._host = host
        self._connect_timeout = connect_timeout
        self._session_timeout = session_timeout
        self._reconnect_timeout = reconnect_timeout
        self._cv = threading.Condition()
        self._handle = None
        if not debug:
            zookeeper.set_log_stream(open("/dev/null"))

    def _handle_connection(self, handle, typ, state, path):
        # The Zookeeper API runs this in a separate event thread
        with self._cv:
            if typ == zookeeper.SESSION_EVENT:
                if state == zookeeper.CONNECTED_STATE:
                    self.connected = True
                elif state == zookeeper.CONNECTING_STATE:
                    self.connected = False
                elif state in (zookeeper.EXPIRED_SESSION_STATE,
                               zookeeper.AUTH_FAILED_STATE):
                    # Last event for this connection, session is dead,
                    # clean up
                    self._handle = None
                    self.connected = False
            self._cv.notify_all()

    def connect(self):
        """Connect to zookeeper"""
        # Run this with our threading Condition
        self._establish_connection = True
        with self._cv:
            # First, check that we didn't just connect
            if self.connected:
                return

            # See if the client is already attempting a connection
            if self._handle is not None:
                start_time = time.time()
                time_taken = 0
                while time_taken <= self._reconnect_timeout:
                    # Release and wait until we hit our timeout
                    # NOTE: We loop here, because the connection handler
                    # releases for *every state change*, and we only care
                    # about getting connected
                    self._cv.wait(self._reconnect_timeout - time_taken)

                    if self.connected:
                        return

                    # Triggered, if the session is expired, we break out
                    # to let logic continue below
                    if self._handle is None:
                        break
                    time_taken = time.time() - start_time

                if self._handle is not None and not self.connected:
                    raise Exception("Timed out waiting for reconnect.")

            # Either first run, or a prior session was ditched entirely
            self._handle = zookeeper.init(self._host, self._handle_connection,
                                         self._session_timeout)
            self._cv.wait(self._connect_timeout)
            if not self.connected:
                raise Exception("Unable to connect to Zookeeper")

    def __getattr__(self, name):
        """Returns a reconnecting version that also uses the current handle"""
        zoo_func = getattr(zookeeper, name)

        def call_func(*args, **kwargs):
            # We wait/try this until we're connected, unless it took
            # too long
            if name == 'close':
                try:
                    return zoo_func(self._handle, *args, **kwargs)
                finally:
                    self._establish_connection = False
                    self._handle = None
                    self.connected = False

            # Check that we're still connected
            if not self.connected and self._establish_connection:
                self.connect()

            start_time = time.time()
            time_taken = 0
            while time_taken <= self._reconnect_timeout:
                try:
                    return zoo_func(self._handle, *args, **kwargs)
                except (zookeeper.ConnectionLossException,
                        zookeeper.SessionExpiredException,
                        zookeeper.SessionMovedException):
                    if not self._establish_connection:
                        raise
                    self.connect()
                except zookeeper.ZooKeeperException, msg:
                    if 'zhandle already freed' in msg and \
                       self._establish_connection:
                        self._handle = None
                        self.connected = False
                        self.connect()
                    else:
                        raise
                time_taken = time.time() - start_time
            raise Exception("Unable to reconnect to execute command.")
        call_func.__doc__ = zoo_func.__doc__

        # Set this function on ourself so that further calls bypass
        # getattr
        setattr(self, name, call_func)
        return call_func

    def __dir__(self):
        return self.__dict__.keys() + zookeeper.__dict__.keys()


zktools/configuration.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is Mozilla zktools.
#
# The Initial Developer of the Original Code is Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2011
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
# Ben Bangert (bbangert@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
"""Zookeeper Nodes

This module provides a :class:`ZkNode` object which can load itself from
a Zookeeper path, and serialize itself back.

"""
import datetime
import decimal
import json
import re
import threading

import zookeeper

ZOO_OPEN_ACL_UNSAFE = {"perms": 0x1f, "scheme": "world", "id": "anyone"}


CONVERSIONS = {
    re.compile(r'^\d+\.\d+$'): decimal.Decimal,
    re.compile(r'^\d+$'): int,
    re.compile(r'^true$', re.IGNORECASE): lambda x: True,
    re.compile(r'^false$', re.IGNORECASE): lambda x: False,
    re.compile(r'^None$', re.IGNORECASE): lambda x: None,
    re.compile(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z$'):
       lambda x: datetime.datetime.strptime(x, '%Y-%m-%dT%H:%M:%S.%fZ'),
    re.compile(r'^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+Z$'):
       lambda x: datetime.datetime.strptime(x, '%Y-%m-%d %H:%M:%S.%fZ'),
    re.compile(r'^\d{4}-\d{2}-\d{2}$'):
       lambda x: datetime.datetime.strptime(x, '%Y-%m-%d'),
}

JSON_REGEX = re.compile(r'^[\{\[].*[\}\]]$')


def _load_value(value, use_json=False):
    """Convert a saved value to the best Python match"""
    for regex, convert in CONVERSIONS.iteritems():
        if regex.match(value):
            return convert(value)
    if use_json and JSON_REGEX.match(value):
        try:
            return json.loads(value)
        except ValueError:
            return value
    return value


def _save_value(value, use_json=False):
    """Convert a Python object to the best string repr"""
    # Float is all we care about, as we lose float precision
    # when calling str on it
    if isinstance(value, float):
        return repr(value)
    elif isinstance(value, datetime.datetime):
        return value.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
    elif isinstance(value, datetime.date):
        return value.strftime('%Y-%m-%d')
    elif use_json and isinstance(value, (dict, list)):
        return json.dumps(value)
    else:
        return str(value)


class ZkNode(object):
    """Zookeeper Node

    This object provides access to a single node for updating the
    value that can also track changes to the value from Zookeeper.

    The value of the node is coerced into an appropriate Python
    object when loaded, current supported conversions::

        Numbers with decimals      -> Decimal
        Numbers without decimals   -> Int
        true/false                 -> Bool
        none                       -> None
        ISO 8601 Date and Datetime -> date or datetime

    And optionally, with use_json::

        JSON string                -> dict/list

    .. note::

        The JSON determination is extremely lax, if its a string that
        starts and ends with brackets or curley marks, its assumed to
        be a JSON object and will be coerced if possible. If coercion
        fails, the string will be returned as is.

    Example::

        from zktools.connection import ZkConnection
        from zktools.configuration import ZkNode

        conn = ZkConnection()
        node = ZkNode(conn, '/some/config/node')

        # prints out the current value, defaults to None
        print node.value

        # Set the value in zookeeper
        node.value = 483.24


    The default behavior is to track changes to the node, so that
    the ``value`` attribute always reflects the node's value in
    Zookeeper.

    .. warning::

        **Do not delete nodes that are in use**, there is purposely
        no code to handle such situations as it creates overly complex
        scenarios both for ZkNode to handle, and for application code
        using it to deal with.

    """
    def __init__(self, connection, path, default=None, use_json=False):
        """Create a Zookeeper Node

        Creating a ZkNode by default attempts to load the value, and
        if its not found will automatically create a blank string as
        the value.

        In the event the node is deleted once this object is deleted
        once its being tracked, the ``deleted`` attribute will be
        ``True``.


        :param connection: zookeeper connection object
        :type connection: ZkConnection instance
        :param path: Path to the Zookeeper node
        :type path: str
        :param default: A default value if the node is being created
        :param use_json: Whether values that look like a JSON object should
                         be deserialized, and dicts/lists saved as JSON.
        :type use_json: bool

        """
        self._zk = connection
        self._path = path
        self._cv = threading.Condition()
        self._use_json = use_json
        self._value = None
        self._reload = False

        with self._cv:
            if not connection.exists(path, self._created_watcher):
                self._zk.create(self._path,
                                _save_value(default, use_json=use_json),
                                [ZOO_OPEN_ACL_UNSAFE], 0)

                # Wait for the node to actually be created
                self._cv.wait()
            self._load()

    def _created_watcher(self, handle, type, state, path):
        """Watch for our node to be created before continuing"""
        with self._cv:
            if type == zookeeper.CREATED_EVENT:
                self._cv.notify_all()

    def _node_watcher(self, handle, type, state, path):
        """Watch a node for updates"""
        with self._cv:
            if type == zookeeper.CHANGED_EVENT:
                data = self._zk.get(self._path, self._node_watcher)[0]
                self._value = _load_value(data, use_json=self._use_json)
            elif type in (zookeeper.EXPIRED_SESSION_STATE,
                          zookeeper.AUTH_FAILED_STATE):
                self._reload = True
            self._cv.notify_all()

    def _load(self):
        """Load data from the node, and coerce as necessary"""
        with self._cv:
            data = self._zk.get(self._path, self._node_watcher)[0]
            self._value = _load_value(data, use_json=self._use_json)

    @property
    def value(self):
        """Returns the current value

        If the Zookeeper session expired, it will be reconnected and
        the value reloaded.

        """
        if self._reload:
            self._load()
            self._reload = False
        return self._value

    @value.setter
    def value(self, value):
        """Set the value with a new one

        :param value: The value of the node
        :type value: Any str'able object

        """
        with self._cv:
            self._zk.set(
                self._path, _save_value(value, use_json=self._use_json))

            # Now wait to see that it triggered our change event
            self._cv.wait()
